modules/smtp/hm-smtp.php

```diff
@@ -315,86 +328,191 @@ function choose_auth() {
         }
         return trim($this->supports_auth[0]);
     }
-
-    /**
-     * authenticate the username and password to the server
-     */
+     // hashes array for scram authentication login case
+    private $hashes = array(
+        'sha-1' => 'sha1',
+        'sha1' => 'sha1',
+        'sha-224' => 'sha224',
+        'sha224' => 'sha224',
+        'sha-256' => 'sha256',
+        'sha256' => 'sha256',
+        'sha-384' => 'sha384',
+        'sha384' => 'sha384',
+        'sha-512' => 'sha512',
+        'sha512' => 'sha512'
+    );
+    
+    private function getHashAlgorithm($scramAlgorithm) {
+        $parts = explode('-', strtolower($scramAlgorithm));
+        if (isset($this->hashes[$parts[1]])) {
+            return $this->hashes[$parts[1]];
+        }
+        return 'sha1'; // Default to sha1 if the algorithm is not found
+    }
+    
+    private function generateClientProof($username, $password, $salt, $clientNonce, $serverNonce, $algorithm) {
+        $iterations = 4096;
+        $keyLength = strlen(hash($algorithm, '', true)); // Dynamically determine key length based on algorithm
+        
+        $passwordBytes = hash($algorithm, $password, true);
+        $saltedPassword = hash_pbkdf2($algorithm, $passwordBytes, $salt, $iterations, $keyLength, true);
+        $clientKey = hash_hmac($algorithm, "Client Key", $saltedPassword, true);
+        $storedKey = hash($algorithm, $clientKey, true);
+        $authMessage = 'n='.$username.',r='.$clientNonce.',s='.$serverNonce;
+        $clientSignature = hash_hmac($algorithm, $authMessage, $storedKey, true);
+        $clientProof = base64_encode($clientKey ^ $clientSignature);
+        return $clientProof;
+    }
```

Comment:
There is a lot of code that repeats in hm-imap and hm-smtp - can you extract to its own module/class and re-use?
I believe you can reuse even the authenticateScram method and abstract the imap or smtp-specific stuff. Consider using a trait or a separate class to handle the Scram authentication as that's pretty heavy otherwise.

Reply:
alright